src/Buffer.mo

```diff
@@ -1,166 +1,1754 @@
-        var i = 0;
-        label l loop {
-          if (i >= count) break l;
-          elems2[i] := elems[i];
-          i += 1;
-        };
-        elems := elems2;
+      if (count == elements.size()) {
+        let elementsSize = elements.size();
+        resize(newUpsize elementsSize);
       };
-      elems[count] := elem;
+      elements[count] := ?elem;
       count += 1;
     };
 
-    /// Removes the item that was inserted last and returns it or `null` if no
-    /// elements had been added to the Buffer.
+    /// Removes and returns the last item in the buffer or `null` if
+    /// the buffer is empty.
+    ///
+    /// Amortized Runtime: O(1), Worst Case Runtime: O(size)
+    ///
+    /// Amortized Space: O(1), Worst Case Space: O(size)
     public func removeLast() : ?X {
       if (count == 0) {
-        null
-      } else {
-        count -= 1;
-        ?elems[count]
+        return null;
+      };
+
+      count -= 1;
+      let lastElement = elements[count];
+      elements[count] := null;
+
+      if (count < elements.size() / DOWNSIZE_THRESHOLD) {
+        resize(elements.size() / DOWNSIZE_FACTOR)
       };
+
+      lastElement
     };
 
-    /// Adds all elements in buffer `b` to this buffer.
-    public func append(b : Buffer<X>) {
-      let i = b.vals();
-      loop {
-        switch (i.next()) {
-          case null return;
-          case (?x) { add(x) };
+    /// Removes and returns the element at `index` from the buffer.
+    /// All elements with index > `index` are shifted one position to the left.
+    /// This may cause a downsizing of the array.
+    /// 
+    /// Runtime: O(size)
+    ///
+    /// Amortized Space: O(1), Worst Case Space: O(size)
+    public func remove(index : Nat) : ?X {
+      if (index >= count) {
+        return null;
+      };
+
+      let element = elements[index];
+
+      // resize down and shift over in one pass
+      if ((count - 1) : Nat < elements.size() / DOWNSIZE_THRESHOLD) { 
+        let elements2 = Prim.Array_init<?X>(elements.size() / DOWNSIZE_FACTOR, null);
+
+        var i = 0;
+        var j = 0;
+        label l while (i < count) {
+          if (i == index) {
+            i += 1;
+            continue l;
+          };
+
+          elements2[j] := elements[i];
+          i += 1;
+          j += 1;
         };
+        elements := elements2;
+      } else { // just shift over elements
+        var i = index;
+        while (i < (count - 1 : Nat)) {
+          elements[i] := elements[i + 1];
+          i += 1;
+        };
+        elements[count] := null;
       };
+
+      count -= 1;
+      element
     };
 
-    /// Returns the current number of elements.
-    public func size() : Nat =
-      count;
+    /// Removes all elements from the buffer for which the predicate returns false.
+    /// The predicate is given both the index of the element and the element itself.
+    /// This may cause a downsizing of the array.
```

Comment:
filterEntries is in place. Because map/mapFilter maps the elements in the buffer to a different type, I don't think there is a good way to write them in-place.